File: ReusableScrollView_iOS_Example/Objc/ReusableScrollView_iOS_Obj-c_Example/ViewController.m
//
//  ViewController.m
//  ReusableScrollView_iOS_Obj-c_Example
//
//  Created by Piotr Domowicz on 15.05.18.
//  Copyright © 2018 sumofighter666. All rights reserved.
//

@import ReusableScrollView;

#import "ViewController.h"
#import "ReusableScrollView.h"

@interface ViewController () <ReusableScrollViewDelegate, ReusableScrollViewDataSource>

@property (weak, nonatomic) IBOutlet ReusableScrollView *scrollView;
@property (nonatomic, assign) NSUInteger viewCount;
@property (nonatomic, assign) CGSize size;

@end

@implementation ViewController

@synthesize initialIndex;
@synthesize numberOfViews;

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    
    self.viewCount = 12;
    
    CGFloat contentWidth = self.size.width * (CGFloat)self.viewCount;
    
    self.scrollView.contentSize         = CGSizeMake(contentWidth, self.size.height);
    self.scrollView.layer.borderColor   = [UIColor blackColor].CGColor;
    self.scrollView.layer.borderWidth   = 1;
    
}

#pragma mark - Getters

- (CGSize)size {
    return self.scrollView.bounds.size;
}

- (NSUInteger)numberOfViews {
    return 12;
}

- (NSInteger)initialIndex {
    return 7;
}
    
- (NSTimeInterval)focusDelay {
    return 0.5;
}

- (UIView *)scrollViewDidRequestViewWithReusableScrollView:(ReusableScrollView *)reusableScrollView atIndex:(NSInteger)atIndex {
    
    // In this case first check the reusable view exists already
    ReusableView *reusableView = [reusableScrollView reusableViewAtAbsoluteIndex:atIndex];
    if ([reusableView.contentView isKindOfClass:[UIImageView class]]) {
        [((UIImageView *)reusableView.contentView) setImage:[self thumbForIndex:atIndex]];
        return nil;
    }
    
    UIImageView *imageView = [[UIImageView alloc] initWithImage:[self thumbForIndex:atIndex]];
    imageView.contentMode = UIViewContentModeScaleAspectFit;
    
    return imageView;
}

- (UIImage *)thumbForIndex:(NSUInteger)index {
    NSString *imageName = [NSString stringWithFormat:@"image%ld_small", index];
    
    return [UIImage imageNamed:imageName];
}

- (UIImage *)imageForIndex:(NSUInteger)index {
    NSString *imageName = [NSString stringWithFormat:@"image%ld", index];
    
    return [UIImage imageNamed:imageName];
}

- (void)reusableViewDidFocusWithReusableView:(ReusableView * _Nonnull)reusableView {
    if (reusableView.absoluteIndex > -1) {
        NSLog(@"Load large image");
        UIImage *image = [self imageForIndex:reusableView.absoluteIndex];
        [((UIImageView *)reusableView.contentView) setImage:image];
    }
}

@end

File: ReusableScrollView_iOS_Example/Swift/ReusableScrollView_iOS_Swift_Example/ViewController.swift
//
//  ViewController.swift
//  ImageScroller
//
//  Created by sumofighter666 on 25.04.18.
//  Copyright © 2018 sumofighter666. All rights reserved.
//

import UIKit
import ReusableScrollView

class ViewController: UIViewController {
    
    
    @IBOutlet weak var scrollView: ReusableScrollView!
    
    lazy private var _size:CGSize = {
        return scrollView.bounds.size
    }()
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        let contentWidth = _size.width * CGFloat(numberOfViews)
        
        scrollView.contentSize = CGSize(width: contentWidth, height: _size.height)
        scrollView.layer.borderColor = UIColor.black.cgColor
        scrollView.layer.borderWidth = 1
    }
    
}

extension ViewController: ReusableScrollViewDelegate, ReusableScrollViewDataSource {
    
    var focusDelay:TimeInterval {
        return 0.5
    }
    
    var initialIndex: Int {
        return 7
    }
    
    var numberOfViews: UInt {
        return 12
    }
    
    func reusableViewDidFocus(reusableView:ReusableView) -> Void {
        guard reusableView.absoluteIndex >= 0 else {
            return
        }
        
        guard let contentView = reusableView.contentView as? UIImageView else {
            return
        }
        
        contentView.image = image(at: UInt(reusableView.absoluteIndex))
    }
    
    func scrollViewDidRequestView(reusableScrollView: ReusableScrollView, atIndex: Int) -> UIView {
        // In this case first check the reusable view exists already
        let reusableView = reusableScrollView.reusableView(atAbsoluteIndex: atIndex)
        
        // Confirm the content view is type of `UIimageView`
        guard
            let contentView = reusableView?.contentView as? UIImageView
            else {
                
                let thumbs = thumb(at: UInt(atIndex))
                let imageView = UIImageView(image: thumbs)
                imageView.contentMode = .scaleAspectFit
                
                return imageView;
        }
        
        contentView.image = thumb(at: UInt(reusableView!.absoluteIndex))
        return contentView;
        
    }
}

extension ViewController {
    func image(at index:UInt) -> UIImage? {
        return UIImage(named: "image\(index)")
    }
    
    func thumb(at index:UInt) -> UIImage? {
        return UIImage(named: "image\(index)_small")
    }
}

